### PR TITLE
Closes #228 | Add M3LS dataloader

### LINKED ISSUE #228: Create dataset loader for M3LS
Dataloader name: `m3ls/m3ls.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?m3ls

|     Dataset| m3ls |
|-------------|---|
| Description | The multilingual multimodal summarization dataset (M3LS) consists of over a million instances of document-image pairs along with a professionally annotated multimodal summary for each pair. It is derived from news articles published by the British Broadcasting Corporation (BBC) over a decade and spans 20 total languages. |
| Subsets | - |
| Languages | ind |
| Tasks | Summarization |
| License | MIT (mit) |
| Homepage | https://github.com/anubhav-jangra/M3LS |
| HF URL | - |
| Paper URL | https://aclanthology.org/2023.eacl-main.263/ |

### PR DESCRIPTION
**Pull request by sabilmakbar**
Please name your PR title and the first line of PR message after the issue it will close. You can use the following examples:
Closes #228 

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/{my_dataset}/{my_dataset}.py` (please use only lowercase and underscore for dataset folder naming, as mentioned in dataset issue) and its `__init__.py` within `{my_dataset}` folder.
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_LOCAL`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py` or `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py --subset_id {subset_name_without_source_or_seacrowd_suffix}`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

**Comment by sabilmakbar** [comment]
Heads-up for reviewers:
1. Total data is ~14GB
2. The `_generate_examples` requires file reconstruction and validation (due to the nature of scraped data requires more validation), in my machine it only generate around 3-4 examples/s and has ~57K examples in total

**Comment by sabilmakbar** [comment]
*pushed-force to remove secrets in earlier commit*

**Review by ljvmiranda921 — APPROVED** [review]
LGTM!

**Comment by holylovenia** [comment]
Hi @patrickamadeus, I would like to let you know that we plan to finalize the calculation of the open contributions (e.g., dataloader implementations) **in 31 hours**, so it'd be great if we could wrap up the reviewing and merge this PR before then.
 
cc: @sabilmakbar

File: seacrowd/sea_datasets/m3ls/m3ls.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+_URL = "https://drive.google.com/uc?id=1Kznkw7YpRiWpdgH4_SVNwp0uGf3j-5e2"
+
+_SUPPORTED_TASKS = [Tasks.SUMMARIZATION, Tasks.IMAGE_CAPTIONING]
+_SOURCE_VERSION = "1.0.0"
+_SEACROWD_VERSION = "1.0.0"
+
+_CONFIG_SUFFIXES_FOR_TASK = [TASK_TO_SCHEMA.get(task).lower() for task in _SUPPORTED_TASKS]
+
+
+class M3LS(datasets.GeneratorBasedBuilder):
```
[the remark this answers is not in the document]
Comment (sabilmakbar) [reply seacrowd/sea_datasets/m3ls/m3ls.py]: Thanks LJ for spotting this!

File: seacrowd/sea_datasets/m3ls/m3ls.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    def _split_generators(self, dl_manager: DownloadManager) -> List[datasets.SplitGenerator]:
+        try:
+            import gdown
+        except ImportError:
+            raise ImportError("Please install `gdown` to enable downloading data from google drive.")
+
+        # Download from Google drive
+        output_dir = os.path.join(os.getcwd(), "data", "m3ls")
+        if not os.path.exists(output_dir):
+            os.mkdir(output_dir)
```
Comment (fhudi) [inline seacrowd/sea_datasets/m3ls/m3ls.py:131]: you should use `os.makedirs()` here.
I received an error due to this.
image
error reference: https://docs.python.org/3/library/os.html#os.mkdir

---

If left as is, this PR is not good for merging.

File: seacrowd/sea_datasets/m3ls/m3ls.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+            subset_id=_DATASETNAME)
+    ] + [
+        SEACrowdConfig(
+            name=f"{_DATASETNAME}_seacrowd_{cfg_sufix}",
+            version=datasets.Version(_SEACROWD_VERSION),
+            description=f"{_DATASETNAME} seacrowd schema for {task.name}",
+            schema=f"seacrowd_{cfg_sufix}",
+            subset_id=_DATASETNAME)
+        for task, cfg_sufix in zip(_SUPPORTED_TASKS, _CONFIG_SUFFIXES_FOR_TASK)
+    ]
```
Comment (fhudi) [inline seacrowd/sea_datasets/m3ls/m3ls.py:81]: This part does not pass the `black` formatter.
Use the following.

```suggestion
    BUILDER_CONFIGS = [
        SEACrowdConfig(
            name=f"{_DATASETNAME}_source",
            version=datasets.Version(_SOURCE_VERSION),
            description=f"{_DATASETNAME} source schema",
            schema="source",
            subset_id=f"{_DATASETNAME}",
        ),
        *[
            SEACrowdConfig(
                name=f"{_DATASETNAME}_seacrowd_{cfg_sufix}",
                version=datasets.Version(_SEACROWD_VERSION),
                description=f"{_DATASETNAME} seacrowd schema for {task.name}",
                schema=f"seacrowd_{cfg_sufix}",
                subset_id=f"{_DATASETNAME}",
            )
            for task, cfg_sufix in zip(_SUPPORTED_TASKS, _CONFIG_SUFFIXES_FOR_TASK)
        ],
    ]
```

File: seacrowd/sea_datasets/m3ls/m3ls.py (hunk 2 above)
Comment (fhudi) [reply seacrowd/sea_datasets/m3ls/m3ls.py:131]: @ljvmiranda921 
Could you help recheck as well?
It seems that you ran the test with additional folder `data` dangling ..

File: seacrowd/sea_datasets/m3ls/m3ls.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+                    "related": datasets.Sequence(datasets.Value("string")),
+                    "section_headers": datasets.Sequence(datasets.Value("string")),
+                    "paragraphs": datasets.Sequence(datasets.Value("string")),
+                    "images": datasets.Sequence(datasets.Image()),
+                    "captions": datasets.Sequence(datasets.Value("string")),
+                }
+            )
+
+        # speech-text schema
+        elif _config_schema_name == f"seacrowd_t2t":
```
Comment (fhudi) [inline seacrowd/sea_datasets/m3ls/m3ls.py:105]: Does not pass `black` formatter
```suggestion
        elif _config_schema_name == "seacrowd_t2t":
```

File: seacrowd/sea_datasets/m3ls/m3ls.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+            features = schemas.text2text_features
+
+        elif _config_schema_name == f"seacrowd_imtext":
```
Comment (fhudi) [inline seacrowd/sea_datasets/m3ls/m3ls.py:108]: Does not pass `black` formatter
```suggestion
        elif _config_schema_name == "seacrowd_imtext":
```

File: seacrowd/sea_datasets/m3ls/m3ls.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+        # the original paper mentioned 80:10:10 splits for over, but there is no info for such splitting index on the extracted folder
+        article_data_dir = os.path.join(local_path, "articles")
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                gen_kwargs={
+                    "article_data_dir": article_data_dir,
+                    "image_folder": os.path.join(local_path, "imagefolder")
+                })
```
Comment (fhudi) [inline seacrowd/sea_datasets/m3ls/m3ls.py:156]:
```suggestion
                    "image_folder": os.path.join(local_path, "imagefolder"),
                },
            )
```

File: seacrowd/sea_datasets/m3ls/m3ls.py [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+                }
+
+            elif _config_schema_name == "seacrowd_t2t":
+                content_data = self.__m3ls_content_data_reconstructor_and_validator(content_data, mode="text")
+                yield idx, {
+                    "id": idx,
+                    "text_1": "\n".join(content_data["paragraphs"]),
+                    "text_2": root_data["summary"],
+                    "text_1_name": "texts",
+                    "text_2_name": "summary"
```
Comment (fhudi) [inline seacrowd/sea_datasets/m3ls/m3ls.py:194]:
```suggestion
                    "text_2_name": "summary",
```

File: seacrowd/sea_datasets/m3ls/m3ls.py [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+        default_values = ([["", ""]], [], "")
+
+        _all_ftr_validation_info = {all_content_ftrs[_idx]: {"dtype": expected_dtypes[_idx], "default_val": default_values[_idx]} for _idx in range(len(all_content_ftrs))}
+
+        if mode == "all":
+            ftr_idx = list(range(3))
+        elif mode == "image":
+            ftr_idx = list(range(1))
+        elif mode == "text":
+            ftr_idx = list(range(1,3))
```
Comment (fhudi) [inline seacrowd/sea_datasets/m3ls/m3ls.py:304]:
```suggestion
            ftr_idx = list(range(1, 3))
```

File: seacrowd/sea_datasets/m3ls/m3ls.py [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+        content_data = content_data_reconstructor(json_content_data)
+
+        content_data_validator(content_data)
+
+        content_data = m3ls_content_data_post_process(content_data)
+
+        return content_data
+
+    @staticmethod
+    def __m3ls_filter_image_and_captions_data(image_data: list, captions_data: list, base_image_folder: str, all_images: list, both_exists: bool=False) -> Tuple[List, List]:
```
Comment (fhudi) [inline seacrowd/sea_datasets/m3ls/m3ls.py:379]:
```suggestion
    def __m3ls_filter_image_and_captions_data(image_data: list, captions_data: list, base_image_folder: str, all_images: list, both_exists: bool = False) -> Tuple[List, List]:
```

**Review by fhudi — CHANGES_REQUESTED** [review]

File: seacrowd/sea_datasets/m3ls/m3ls.py (hunk 2 above)
Comment (sabilmakbar) [reply seacrowd/sea_datasets/m3ls/m3ls.py:131]: thanks! yeah previously I had the folder `data` exists

Comment (fhudi) [reply seacrowd/sea_datasets/m3ls/m3ls.py:131]:
```suggestion
        output_dir = os.path.join(os.getcwd(), "data", "m3ls")
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
```

**Comment by sabilmakbar** [comment]
done making changes, pls have a look @fhudi

**Comment by fhudi** [comment]
@sabilmakbar cool. thanks.
The test is currently running, since it is a 15.9GB file, please wait for a while

**Review by fhudi — CHANGES_REQUESTED** [review]
@sabilmakbar 
Pillow library is required.
Shall we add try-except for importing?
image

**Comment by fhudi** [comment]
@sabilmakbar 
Other than the above mentioned issue, everything else work fine.
Passed the test and reviewing check-list. 🙏

**Comment by sabilmakbar** [comment]
> Pillow library is required.
> Shall we add try-except for checking?

Hi @fhudi, I thought `PIL` was included in SEACrowd reqs (but in fact it isn't). I'll add the try-catch exception for now so that every user can avoid downloading it only to find the error when generating the dataset due to missing `PIL` lib. Thanks for pointing this

**Pull request merged**